internal/backend/kubernetes/config.go

```diff
@@ -1,15 +1,26 @@
 package kubernetes
 
 import (
+	"context"
+
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
 	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 )
 
 // Config used by the NewBackend function family.
 type Config struct {
+	// Config is the Kubernetes client configuration.
 	*rest.Config
+
+	// Namespace restricts the scope of the backend such that Hardware objects are retrieved from
+	// this namespace only. Defaults to "default".
 	Namespace string
+
+	// Context is the context used by the Kubernetes client. Defaults to context.Background().
+	// When specified it controls the lifetime of the Kubernetes client by shutting the client
+	// down when it cancelled.
+	Context context.Context
```

Comment:
curious about your thought process here. context in a struct vs passing the same context around to all func/methods?

Reply:
The struct is just a configuration object. It doesn't do anything but provide data to the `NewBackend`. `NewBackend` doesn't store the whole context, it only stores the `Done()` channel. This is for convenience and to avoid a `Start()` method because, in times gone by, I've found that nearly always results in lots of `NewBackend()` _immediately_ followed by `Start()`. I could go either way though.

If we consider this from the consumers point of view:


```go
func main() {
	ctx, cancel := signals.NotifyContext(context.Background, syscall.SIGUP)
	defer cancel()

	// This
	be := kubernetes.NewBackend(kubernetes.Config{
		// Other config
		Context: ctx,
	})

	// vs this
	be := kubernetes.NewBackend(kubernetes.Config{
		// Other config
	})
	be.Start(ctx)
}